archive/import.go

```diff
@@ -2,156 +2,344 @@ package archive
+	zctx := resolver.NewContext()
+	indexBuilder := zng.NewBuilder(zctx.MustLookupTypeRecord([]zng.Column{
+		{"ts", zng.TypeTime},
+		{"offset", zng.TypeInt64},
+	}))
+	return &chunkWriter{
+		dataFile:        dataFile,
+		dataFileWriter:  dataFileWriter,
+		indexBuilder:    indexBuilder,
+		indexTempPath:   indexTempPath,
+		indexTempWriter: indexTempWriter,
+		tsDir:           tsDir,
+		needIndexWrite:  true,
+	}, nil
+}
+
+func (c *chunkWriter) Write(rec *zng.Record) error {
 	// We want to index the start of stream (SOS) position of the data file by
 	// record timestamps; we don't know when we've started a new stream until
 	// after we written the first record in the stream.
-	sos := d.dataFileWriter.LastSOS()
-	if err := d.dataFileWriter.Write(rec); err != nil {
+	sos := c.dataFileWriter.LastSOS()
+	if err := c.dataFileWriter.Write(rec); err != nil {
 		return err
 	}
-	d.lastTs = rec.Ts()
-	d.rcount++
-	if d.needIndexWrite {
-		out := d.indexBuilder.Build(zng.EncodeTime(d.lastTs), zng.EncodeInt(sos))
-		if err := d.indexTempWriter.Write(out); err != nil {
-			return err
-		}
-		d.needIndexWrite = false
+	c.lastTs = rec.Ts()
+	if c.firstTs == 0 {
+		c.firstTs = c.lastTs
 	}
-	if d.dataFileWriter.LastSOS() != sos {
-		d.needIndexWrite = true
-	}
-	if d.dataFileWriter.Position() >= d.ark.LogSizeThreshold {
-		if err := d.close(); err != nil {
+	c.rcount++
+	if c.needIndexWrite {
+		out := c.indexBuilder.Build(zng.EncodeTime(c.lastTs), zng.EncodeInt(sos))
+		if err := c.indexTempWriter.Write(out); err != nil {
 			return err
 		}
+		c.needIndexWrite = false
+	}
+	if c.dataFileWriter.LastSOS() != sos {
+		c.needIndexWrite = true
 	}
 	return nil
 }
 
-func (d *importDriver) close() error {
-	if d.dataFileWriter == nil {
-		return nil
-	}
-	err := d.dataFileWriter.Close()
-	if closeErr := d.indexTempWriter.Close(); err == nil {
+func (c *chunkWriter) close(ctx context.Context) error {
+	err := c.dataFileWriter.Close()
+	if closeErr := c.indexTempWriter.Close(); err == nil {
 		err = closeErr
 	}
 	if err != nil {
 		return err
 	}
-	d.dataFileWriter = nil
+	c.dataFileWriter = nil
 	// Write the time seek index into the archive, feeding it the key/offset
 	// records written to indexTempPath.
-	tf, err := fs.Open(d.indexTempPath)
+	tf, err := fs.Open(c.indexTempPath)
 	if err != nil {
 		return err
 	}
 	defer func() {
 		tf.Close()
 		os.Remove(tf.Name())
 	}()
-	tfr := zngio.NewReader(tf, d.zctx)
-	sf := seekIndexFile{id: d.dataFile.id, recordCount: d.rcount, first: d.firstTs, last: d.lastTs}
-	idxURI := d.tsDir.AppendPath(sf.name())
-	idxWriter, err := microindex.NewWriter(d.zctx, idxURI.String(), []string{"ts"}, framesize)
+	tfr := zngio.NewReader(tf, resolver.NewContext())
+	sf := seekIndexFile{id: c.dataFile.id, recordCount: c.rcount, first: c.firstTs, last: c.lastTs}
+	idxURI := c.tsDir.AppendPath(sf.name())
+	idxWriter, err := microindex.NewWriter(resolver.NewContext(), idxURI.String(), []string{"ts"}, framesize)
```

Comment:
```suggestion
	idxWriter, err := microindex.NewWriter(zctx, idxURI.String(), []string{"ts"}, framesize)
```

Reply:
Out of curiosity, why do they need the same context here?